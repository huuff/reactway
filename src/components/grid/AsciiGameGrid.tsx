
import { Grid } from "../../grid/grid";
import { Fragment } from "react";

type GameGridProps = {
    grid: Grid
} & { className: string };

const AsciiGameGrid = ({ grid, className }: GameGridProps) => (
    <div className={`${className} font-mono leading-none text-lg`}>
        {[...Array(grid.height)].map((_, y) => (
            <Fragment key={`row-${y}`}>
                {
                    [...Array(grid.width)].map((_, x) => (
                        <span key={`${x}-${y}`}>{grid.get(x, y) ? "X" : "O"}</span>
                    ))
                }
                <br key={`row-break-${y}`} />
            </Fragment>
        ))}
    </div>
)

export default AsciiGameGrid;