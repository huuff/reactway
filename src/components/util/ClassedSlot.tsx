import isArray from "lodash/isArray";
import { Children, cloneElement, ReactElement } from "react";

type ClassedSlotProps = {
    children: ReactElement | ReactElement[];
    className: string,
}

function withClassName(element: ReactElement, className: string): ReactElement {
    return cloneElement(
            element,
            { ...element.props, className: `${element.props.className} ${className}`}
        );   
}
const ClassedSlot = ({ children, className }: ClassedSlotProps) => {
    return (
        <>
            {
                isArray(children)
                    ? Children.map(children, (child) => withClassName(child, className))
                    : withClassName(children, className)
            }
        </>
    )
}


export default ClassedSlot;