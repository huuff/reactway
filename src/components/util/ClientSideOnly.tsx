import dynamic from 'next/dynamic';
import React, { ReactNode } from 'react';

type NoSsrProps = {
    children: ReactNode
}

const ClientSideOnly = (props: NoSsrProps) => (
  <React.Fragment>{props.children}</React.Fragment>
);

export default dynamic(() => Promise.resolve(ClientSideOnly), {
  ssr: false
});
