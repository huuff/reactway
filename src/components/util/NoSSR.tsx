import dynamic from 'next/dynamic';
import React, { ReactNode } from 'react';

type NoSsrProps = {
    children: ReactNode
}

const NoSsr = (props: NoSsrProps) => (
  <React.Fragment>{props.children}</React.Fragment>
);

export default dynamic(() => Promise.resolve(NoSsr), {
  ssr: false
});
