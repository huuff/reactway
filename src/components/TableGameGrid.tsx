
import { Grid } from "../grid/grid";
import cx from "classnames";

type GameGridProps = {
    grid: Grid<any>
} & { className: string };

const TableGameGrid = ({ grid, className }: GameGridProps) => (
    <table className={`${className}`}>
        <tbody>
            {[...Array(grid.height)].map((_, y) => (
                <tr key={`row-${y}`}>
                    {
                        [...Array(grid.width)].map((_, x) => (
                            <td
                                key={`${x}-${y}`}
                                className={cx({
                                    "bg-black": grid.get(x, y),
                                    "bg-white": !grid.get(x, y),
                                    "w-6": true,
                                    "h-6": true,
                                    "border": true,
                                })}>
                            </td>
                        ))
                    }
                </tr>
            ))}
        </tbody>
    </table>
)

export default TableGameGrid;