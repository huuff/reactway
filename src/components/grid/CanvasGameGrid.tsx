import { CSSProperties, MouseEventHandler, RefObject, useCallback, useEffect, useMemo, useRef } from "react";
import { Coordinates, GameGridProps, Grid } from "../../grid/grid";
import { useMouseState, useViewportState } from "beautiful-react-hooks";
import { useDarkMode, useDebounce } from "usehooks-ts";
import { Box2D } from "../../util/box-2d";
import tuple from "immutable-tuple";

const CELL_SIZE_MULTIPLIER = 8;

type CanvasGameGridProps = GameGridProps & {
    scrollX?: number;
    scrollY?: number;
};
type Size = {
    width: number;
    height: number;
}

function getBoundingRectOrZeros(ref: RefObject<HTMLElement>) {
    return ref.current?.getBoundingClientRect() ?? { left: 0, top: 0, right: 0, bottom: 0};
}

// TODO: Test it? Can I?
const CanvasGameGrid = ({
    grid, 
    className, 
    toggleCell,
    cellSize,
    scrollX = 0,
    scrollY = 0,
 }: CanvasGameGridProps) => {
    const { isDarkMode } = useDarkMode();
    const cellSizePixels = useMemo(() => CELL_SIZE_MULTIPLIER * cellSize, [cellSize]);
    const gridCanvasRef = useRef<HTMLCanvasElement>(null);

    const { width: windowWidth, height: windowHeight } = useViewportState();
    
    const gridSizePixels = useMemo(() => ({
        width: grid.width * cellSizePixels,
        height: grid.height * cellSizePixels,
    }), [grid.height, grid.width, cellSizePixels]);

    const visibleCellBounds = useVisibleBounds(windowWidth, windowHeight, scrollX, scrollY, gridSizePixels, cellSizePixels)

    useDrawCanvasEffect(gridCanvasRef, grid, cellSizePixels, visibleCellBounds, isDarkMode);

    const hoveredCell = useHoveredCell(grid.width, grid.height, gridCanvasRef, cellSizePixels);
    const onMouseUp = useClickToggleHandler(gridCanvasRef, hoveredCell, grid, toggleCell);

    const highlightedCell = useHighlightedCell(gridCanvasRef, cellSizePixels, hoveredCell);

    const isMouseWithinGrid = useIsMouseWithinGrid(gridCanvasRef);

    return (
        <div onMouseUp={onMouseUp} className={className}>
            { isMouseWithinGrid && (
            <div style={{
                position: "absolute",
                height: `${cellSizePixels}px`,
                width: `${cellSizePixels}px`,
                backgroundColor: "red",
                left: highlightedCell[0],
                top: highlightedCell[1],
                zIndex: 10,
                opacity: 0.5,
            }} />
            )}
            <canvas
                ref={gridCanvasRef}
                className="mx-auto"
                id="grid-canvas"
                height={gridSizePixels.height}
                width={gridSizePixels.width}
            ></canvas>
        </div>
    )
}

function useIsMouseWithinGrid(gridCanvasRef: RefObject<HTMLCanvasElement>): boolean {
    const { clientX, clientY } = useMouseState();
    const boundingRect = gridCanvasRef.current?.getBoundingClientRect();
    if (!boundingRect) {
        return false;
    }

    const { left, top, right, bottom } = boundingRect;

    return clientX >= left && clientX  <= right && clientY >= top && clientY <= bottom;
}

/**
 * Returns a box that indicates the bounds of the visible part of the screen
 */
function useVisibleBounds(
    windowWidth: number,
    windowHeight: number,
    scrollX: number,
    scrollY: number,
    gridSizePixels: Size,
    cellSizePixels: number
): Box2D {
    const visibleBounds = useDebounce(useMemo<Box2D>(() => new Box2D(
        tuple(
            Math.max(scrollX - (windowWidth / 2), 0),
            Math.max(scrollY - (windowHeight / 2), 0)
        ),
        tuple(
            Math.min(scrollX + (windowWidth * 1.5), gridSizePixels.width),
            Math.min(scrollY + (windowHeight * 1.5), gridSizePixels.height),
        ),
    ), [windowHeight, windowWidth, scrollX, scrollY, gridSizePixels]), 250);
    const visibleCellBounds = useMemo<Box2D>(
        () => visibleBounds.divide(cellSizePixels),
        [visibleBounds, cellSizePixels]
    );

    return visibleCellBounds;
}

/**
 * Returns coordinates that match the position of the highlighted cell in the screen,
 * taking into account all displacements and transformations
 */
const useHighlightedCell = (
    gridCanvasRef: RefObject<HTMLCanvasElement>, 
    cellSizePixels: number,
    [hoveredCellX, hoveredCellY]: Coordinates
    ): Coordinates => {

    const { left: leftDisplacement, top } = getBoundingRectOrZeros(gridCanvasRef);

    const topDisplacement = Math.max(top, 0);

    const coordinates = tuple(
        (hoveredCellX * cellSizePixels) + leftDisplacement,
        (hoveredCellY * cellSizePixels) + topDisplacement,
    );

    return coordinates;
}

/**
 * Returns the cell (in the grid data structure) that the cursor is hovering
 */
const useHoveredCell = (
    gridWidth: number,
    gridHeight: number,
    gridCanvasRef: RefObject<HTMLCanvasElement>,
    cellSizePixels: number
): Coordinates => {
    const { clientX, clientY } = useMouseState();
    const { width: viewportWidth, height: viewportHeight } = useViewportState();

    const { left: leftDisplacement, top, } = getBoundingRectOrZeros(gridCanvasRef);

    const gridDisplacementToTheBottom = Math.max(top, 0);

    const mouseX = Math.max(0, clientX - leftDisplacement)
    const mouseY = Math.max(0, clientY - gridDisplacementToTheBottom)

    const result = tuple(
        Math.min(Math.floor((mouseX - (mouseX % cellSizePixels)) / cellSizePixels), gridWidth-1),
        Math.min(Math.floor((mouseY - (mouseY % cellSizePixels)) / cellSizePixels), gridHeight-1),
    );

    return result;
}

function useClickToggleHandler(
    ref: RefObject<HTMLCanvasElement>,
    mouseCell: Coordinates,
    grid: Grid,
    toggleCell: (coordinates: Coordinates) => void,
): MouseEventHandler<HTMLElement> {
    return useCallback((event) => {
        if (grid.contains(mouseCell)) {
            toggleCell(mouseCell);
        }
    }, [ref, mouseCell, grid, toggleCell]);
}

function useDrawCanvasEffect(
    ref: RefObject<HTMLCanvasElement>,
    grid: Grid,
    cellSizePixels: number,
    visibleCellBounds: Box2D,
    isDarkMode: boolean
) {
    useEffect(() => {
        const canvas = ref.current!;

        const ctx = canvas.getContext("2d")!;
        for (const { coordinates: [x, y], isAlive } of grid.boundedIterator(visibleCellBounds)) {
            if (isAlive) {
                ctx.fillStyle = isDarkMode ? "#262626" : "#000000";
            } else {
                ctx.fillStyle = isDarkMode ? "#666666" : "#F0F0F0";
                ctx.strokeRect(x * cellSizePixels, y * cellSizePixels, cellSizePixels, cellSizePixels);
            }
            ctx.fillRect(x * cellSizePixels, y * cellSizePixels, cellSizePixels, cellSizePixels);
        };
    }, [grid, cellSizePixels, visibleCellBounds, isDarkMode])
}

export default CanvasGameGrid;