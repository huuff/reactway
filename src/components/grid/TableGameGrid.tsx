
import { GameGridProps } from "../../grid/grid";
import classNames from "classnames";
import { coordinatesToString } from "../../util/coordinates-to-string";

const TableGameGrid = ({ grid, className, toggleCell }: GameGridProps) => (
    <table className={`${className || ""} w-72 table-fixed`}>
        <tbody>
            {[...Array(grid.height)].map((_, y) => (
                <tr key={`row-${y}`}>
                    {
                        [...Array(grid.width)].map((_, x) => {
                            const coord = coordinatesToString([x, y]);
                            return (
                                <td
                                    key={coord}
                                    data-testid={coord}
                                    onClick={() => toggleCell([x, y])}
                                    className={classNames(
                                        "w-6",
                                        "h-6",
                                        "border",
                                        {
                                            "bg-black": grid.get(x, y),
                                            "bg-white": !grid.get(x, y),
                                            "hover:bg-red-800": grid.get(x, y),
                                            "hover:bg-red-400": !grid.get(x, y),
                                        })}>
                                </td>
                            )
                        })
                    }
                </tr>
            ))}
        </tbody>
    </table>
)

export default TableGameGrid;