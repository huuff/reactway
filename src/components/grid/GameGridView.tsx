import { GameGridProps } from "../../grid/grid";
import { GridViewType } from "../../settings/settings";
import AsciiGameGrid from "./AsciiGameGrid";
import CanvasGameGrid from "./CanvasGameGrid";
import TableGameGrid from "./TableGameGrid";

type GameGridViewProps = GameGridProps & {
    view: GridViewType;
    scrollX: number,
    scrollY: number,
}

const GameGridView = (props: GameGridViewProps) => {
    switch (props.view) {
        case "table":
            return <TableGameGrid {...props} className="mx-auto" />
        case "ascii":
            return <AsciiGameGrid {...props} className="text-center" />
        case "canvas":
            return <CanvasGameGrid {...props} className="mx-auto" />
    }
}

export default GameGridView;