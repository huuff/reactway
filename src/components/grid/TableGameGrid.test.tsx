import TableGameGrid from "./TableGameGrid";
import { render, screen } from "@testing-library/react";
import { SetGrid } from "../../grid/set-grid";
import { coordinatesToString } from "../../util/coordinates-to-string";
import { Coordinates } from "../../grid/grid";
import renderer from "react-test-renderer";
import tuple from "immutable-tuple";

jest.mock("usehooks-ts", () => ({
    ...jest.requireActual("usehooks-ts"),
    useDarkMode: jest.fn(() => ({
      isDarkMode: false,
    }))
  }));

const liveCells: Coordinates[] = [tuple(2, 3), tuple(2, 1), tuple(3, 2), tuple(3,3), tuple(4, 2)];
describe("TableGameGrid", () => {
    test("live cells have the alive color", () => {
        render(<TableGameGrid grid={new SetGrid(liveCells)} toggleCell={jest.fn()} cellSize={3}/>);

        for (const coord of liveCells) {
            expect(screen.getByTestId(coordinatesToString(coord))).toHaveClass("bg-light-alive-cell");
        }
    });

    test("dead cells have the dead color", () => {
        render(<TableGameGrid grid={new SetGrid(liveCells)} toggleCell={jest.fn()} cellSize={3}/>);

        expect(screen.getByTestId(coordinatesToString(tuple(2, 2)))).toHaveClass("bg-light-dead-cell");
    });

    test("snapshot", () => {
        expect(renderer.create(
            <TableGameGrid grid={new SetGrid(liveCells)} toggleCell={jest.fn()} cellSize={3}/>
        ).toJSON()).toMatchSnapshot();
    });
})