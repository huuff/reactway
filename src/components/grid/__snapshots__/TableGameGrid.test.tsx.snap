// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableGameGrid snapshot 1`] = `
<table
  className=" w-72 table-fixed"
>
  <tbody>
    <tr>
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(0,0)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(1,0)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(2,0)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(3,0)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(4,0)"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(0,1)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(1,1)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-black hover:bg-red-800"
        data-testid="(2,1)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(3,1)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(4,1)"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(0,2)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(1,2)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(2,2)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-black hover:bg-red-800"
        data-testid="(3,2)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-black hover:bg-red-800"
        data-testid="(4,2)"
        onClick={[Function]}
      />
    </tr>
    <tr>
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(0,3)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(1,3)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-black hover:bg-red-800"
        data-testid="(2,3)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-black hover:bg-red-800"
        data-testid="(3,3)"
        onClick={[Function]}
      />
      <td
        className="h-NaN w-NaN border bg-white hover:bg-red-400"
        data-testid="(4,3)"
        onClick={[Function]}
      />
    </tr>
  </tbody>
</table>
`;
